spec/models/manageiq/providers/kubernetes/container_manager/metrics_capture_spec.rb

```diff
@@ -0,0 +1,84 @@
+require "spec_helper"
+
+describe ManageIQ::Providers::Kubernetes::ContainerManager::MetricsCapture do
+  before do
+    _guid, _server, @zone = EvmSpecHelper.create_guid_miq_server_zone
+
+    @ems_kubernetes = FactoryGirl.create(:ems_kubernetes, :zone => @zone)
+
+    @node = FactoryGirl.create(:kubernetes_node,
+                               :ext_management_system => @ems_kubernetes,
+                               :ems_ref               => 'target')
+
+    @node.computer_system =  FactoryGirl.create(:computer_system)
+    @node.computer_system.hardware = FactoryGirl.create(
+      :hardware,
+      :cpu_total_cores => 2,
+      :memory_mb       => 2048)
+
+    @group = FactoryGirl.create(:container_group,
+                                :ext_management_system => @ems_kubernetes,
+                                :container_node        => @node)
+
+    @container = FactoryGirl.create(:kubernetes_container,
+                                    :container_group => @group,
+                                    :ems_ref         => 'target')
+  end
+
+  context "#perf_collect_metrics" do
+    it "raises an error when no ems is defined" do
+      @node.ext_management_system = nil
+      expect { @node.perf_collect_metrics('interval_name') }.to raise_error(
+        described_class::TargetValidationError)
+    end
+
+    it "raises an error when no cpu cores are defined" do
+      @node.hardware.cpu_total_cores = nil
+      expect { @node.perf_collect_metrics('interval_name') }.to raise_error(
+        described_class::TargetValidationError)
+    end
+
+    it "raises an error when memory is not defined" do
+      @node.hardware.memory_mb = nil
+      expect { @node.perf_collect_metrics('interval_name') }.to raise_error(
+        described_class::TargetValidationError)
+    end
+
+    # TODO: include also sort_and_normalize in the tests
+    METRICS_EXERCISES = [
+      {:counters => [
+        {'start' => 1446500000000, 'end' => 1446500020000, 'avg' => 0},
+        {'start' => 1446500020000, 'end' => 1446500040000, 'avg' => 4000000000}],
+       :gauges   => [
+        {'start' => 1446500000000, 'end' => 1446500020000, 'avg' => 1073741824}],
+       :expected => {}
+      },
+      {:counters => [
+        {'start' => 1446499980000, 'end' => 1446500020000, 'avg' => 0},
+        {'start' => 1446500000000, 'end' => 1446500040000, 'avg' => 4000000000}],
+       :gauges   => [
+        {'start' => 1446500000000, 'end' => 1446500020000, 'avg' => 1073741824}],
+       :expected => {
+         Time.at(1446500000).utc => {
+           "cpu_usage_rate_average"     => 10.0,
+           "mem_usage_absolute_average" => 50.0
+         }}
+      }
+    ]
+
+    it "node counters and gauges are correctly processed" do
+      [@node, @container].each do |target|
+        METRICS_EXERCISES.each do |exercise|
```

Comment:
This nesting creates one test that tries to test all combinations.  If you get a failure you are going to have a hard time narrowing down.  It might be better to generate separate specs, which will help in narrowing down which one could fail...at the cost of test performance (since it has to re-setup each time)

I'm thinking something like...

```ruby
[@node, @container].each do |target|
  it "counters and gauges are correctly processed for #{target.class}" do
    METRICS_EXERCISES.each do |exercise|
      ...
    end
  end
end
```

In addition, it might be clearer to unwind the METRICS_EXERCISES and call the parts individually.  It is less DRY, but would be much clearer in understanding that the spec is trying to do...

```ruby
# Initial data
counters, gauges, expected = METRICS_EXERCISES[0].values_at(:counters, :gauges, :expected)
described_class::CaptureContext.any_instance.stub(:fetch_counters_data).and_return(counters)
described_class::CaptureContext.any_instance.stub(:fetch_gauges_data).and_return(gauges)

_, values_by_ts = target.perf_collect_metrics('realtime')

expect(values_by_ts['target']).to eq(expected)

# Second set of data
counters, gauges, expected = METRICS_EXERCISES[1].values_at(:counters, :gauges, :expected)
described_class::CaptureContext.any_instance.stub(:fetch_counters_data).and_return(counters)
described_class::CaptureContext.any_instance.stub(:fetch_gauges_data).and_return(gauges)

_, values_by_ts = target.perf_collect_metrics('realtime')

expect(values_by_ts['target']).to eq(expected)
```